File: packages/types/src/domain-config.ts
import { z } from 'zod';

// PR12 §15 — per-organization domain configuration. Shape mirrors
// PRD-Analysis-Pipeline.md §15.1's domain_config.json example.
// Industry-specific behavior (terminology rules, jurisdiction risk,
// risk taxonomy, factual baseline) lives here per-organization
// rather than in code, so the same codebase can serve different
// in-house legal teams.

export const TerminologyRuleSchema = z.object({
  preferred: z.string().min(1),
  avoid: z.string().min(1),
  rationale: z.string().optional(),
});
export type TerminologyRule = z.infer<typeof TerminologyRuleSchema>;

export const VerbRuleSchema = z.object({
  prefer: z.string().min(1),
  avoid: z.string().min(1),
  context: z.string().optional(),
});
export type VerbRule = z.infer<typeof VerbRuleSchema>;

export const HighScrutinyJurisdictionSchema = z.object({
  jurisdiction: z.string().min(1),
  rationale: z.string().optional(),
  appliesToPracticeAreas: z.array(z.string()).default([]),
});
export type HighScrutinyJurisdiction = z.infer<typeof HighScrutinyJurisdictionSchema>;

export const DomainRiskCategorySchema = z.object({
  categoryId: z.string().min(1),
  label: z.string().min(1),
  examplesFlag: z.array(z.string()).default([]),
  defaultSeverity: z.enum(['high', 'medium', 'low']),
});
export type DomainRiskCategory = z.infer<typeof DomainRiskCategorySchema>;

export const EscalationThresholdsSchema = z
  .object({
    financialBetTheCompanyUsd: z.number().nonnegative().optional(),
    confidenceLowRouting: z
      .enum(['senior_reviewer', 'queue', 'requester_only'])
      .optional(),
  })
  .default({});
export type EscalationThresholds = z.infer<typeof EscalationThresholdsSchema>;

export const DomainConfigSchema = z.object({
  // Notion page id (or other KB pointer) that holds the org's
  // standing fact pattern. Skills don't fetch this directly; the
  // worker can pull excerpts into the prompt when relevant.
  factualBaselineKbPage: z.string().optional(),
  // Inline factual baseline facts injected into every research skill's
  // user prompt. Keep short — these are static premises about the org.
  factualBaselineFacts: z.array(z.string()).default([]),
  terminologyRules: z.array(TerminologyRuleSchema).default([]),
  verbRules: z.array(VerbRuleSchema).default([]),
  highScrutinyJurisdictions: z.array(HighScrutinyJurisdictionSchema).default([]),
  domainRiskTaxonomy: z.array(DomainRiskCategorySchema).default([]),
  preferredResearchDatabase: z
    .enum(['westlaw', 'lexis', 'fastcase', 'courtlistener', 'vlex', 'bloomberg', 'mixed'])
    .optional(),
  escalationThresholds: EscalationThresholdsSchema,
});
export type DomainConfig = z.infer<typeof DomainConfigSchema>;

// Empty config returned by the loader when an organization has no
// custom rules. All consumers must handle the empty case identically
// to a missing config — the system is fully functional without any
// per-org customization.
export const EMPTY_DOMAIN_CONFIG: DomainConfig = {
  factualBaselineFacts: [],
  terminologyRules: [],
  verbRules: [],
  highScrutinyJurisdictions: [],
  domainRiskTaxonomy: [],
  escalationThresholds: {},
};

// Returns true if the config has any content the skills should
// surface. Used to short-circuit prompt rendering when the config
// is empty (most pre-customization deployments).
export function hasDomainConfigContent(c: DomainConfig): boolean {
  return (
    c.factualBaselineFacts.length > 0 ||
    c.terminologyRules.length > 0 ||
    c.verbRules.length > 0 ||
    c.highScrutinyJurisdictions.length > 0 ||
    c.domainRiskTaxonomy.length > 0
  );
}

File: apps/ai/src/domain_config.py
"""PR12 §15 — domain-config shape + prompt-rendering helpers.

Mirrors packages/types/src/domain-config.ts. The worker passes a
domain_config dict to every skill request; this module gives the
skills a typed view + a helper that turns the config into a
prompt-ready text block, or an empty string when the config has no
content.
"""

from typing import Literal

from pydantic import BaseModel, Field


class TerminologyRule(BaseModel):
    preferred: str
    avoid: str
    rationale: str | None = None


class VerbRule(BaseModel):
    prefer: str
    avoid: str
    context: str | None = None


class HighScrutinyJurisdiction(BaseModel):
    jurisdiction: str
    rationale: str | None = None
    applies_to_practice_areas: list[str] = Field(default_factory=list)


class DomainRiskCategory(BaseModel):
    category_id: str
    label: str
    examples_flag: list[str] = Field(default_factory=list)
    default_severity: Literal["high", "medium", "low"]


class DomainConfig(BaseModel):
    factual_baseline_facts: list[str] = Field(default_factory=list)
    terminology_rules: list[TerminologyRule] = Field(default_factory=list)
    verb_rules: list[VerbRule] = Field(default_factory=list)
    high_scrutiny_jurisdictions: list[HighScrutinyJurisdiction] = Field(default_factory=list)
    domain_risk_taxonomy: list[DomainRiskCategory] = Field(default_factory=list)


def domain_config_block(cfg: DomainConfig | None) -> str:
    """Render the domain config as a prompt block.

    Returns an empty string when the config has no content — the
    skills can unconditionally append `domain_config_block(...)` to
    their user prompt without polluting the context window when the
    org hasn't customized anything.
    """
    if cfg is None:
        return ""
    has_content = (
        cfg.factual_baseline_facts
        or cfg.terminology_rules
        or cfg.verb_rules
        or cfg.high_scrutiny_jurisdictions
        or cfg.domain_risk_taxonomy
    )
    if not has_content:
        return ""

    parts: list[str] = ["", "--- Organization domain configuration (PRD §15) ---"]

    if cfg.factual_baseline_facts:
        parts.append("Factual baseline (standing facts about the organization):")
        for f in cfg.factual_baseline_facts:
            parts.append(f"  - {f}")

    if cfg.terminology_rules:
        parts.append("")
        parts.append("Terminology rules — use the preferred term; avoid the alternative:")
        for tr in cfg.terminology_rules:
            line = f"  prefer: {tr.preferred!r}  ·  avoid: {tr.avoid!r}"
            if tr.rationale:
                line += f"  ({tr.rationale})"
            parts.append(line)

    if cfg.verb_rules:
        parts.append("")
        parts.append(
            "Verb rules — when describing what the organization does, prefer the first verb; "
            "the second verb may create exposure or misframing:"
        )
        for vr in cfg.verb_rules:
            line = f"  prefer: {vr.prefer!r}  ·  avoid: {vr.avoid!r}"
            if vr.context:
                line += f"  ({vr.context})"
            parts.append(line)

    if cfg.high_scrutiny_jurisdictions:
        parts.append("")
        parts.append(
            "High-scrutiny jurisdictions — extra care for any analysis touching these:"
        )
        for j in cfg.high_scrutiny_jurisdictions:
            tail = ""
            if j.applies_to_practice_areas:
                tail = f"  [areas: {', '.join(j.applies_to_practice_areas)}]"
            parts.append(
                f"  - {j.jurisdiction}: {j.rationale or '(no rationale provided)'}{tail}"
            )

    if cfg.domain_risk_taxonomy:
        parts.append("")
        parts.append(
            "Domain risk taxonomy — every output should classify against these categories. "
            "Flag each as 'risk-flagged' or 'clear' explicitly; do not leave any blank:"
        )
        for c in cfg.domain_risk_taxonomy:
            parts.append(
                f"  - [{c.category_id}] {c.label}  (default severity: {c.default_severity})"
            )

    return "\n".join(parts)
